kubelet/datadog_checks/kubelet/kubelet.py

```diff
@@ -344,6 +352,49 @@ def _report_container_spec_metrics(self, pod_list, instance_tags):
                 except (KeyError, AttributeError) as e:
                     self.log.debug("Unable to retrieve container limits for %s: %s", c_name, e)
 
+    def _report_container_state_metrics(self, pod_list, instance_tags):
+        """Reports container state & reasons by looking at container statuses"""
+        for pod in pod_list['items']:
+            pod_name = pod.get('metadata', {}).get('name')
+            pod_uid = pod.get('metadata', {}).get('uid')
+
+            if not pod_name or not pod_uid:
+                continue
+
+            for ctr_status in pod['status'].get('containerStatuses', []):
+                c_name = ctr_status.get('name')
+                cid = ctr_status.get('containerID')
+                if not c_name or not cid:
+                    continue
+
+                if self.pod_list_utils.is_excluded(cid, pod_uid):
+                    continue
+
+                tags = get_tags('%s' % cid, True) + instance_tags
+
+                restart_count = ctr_status.get('restartCount', 0)
+                self.gauge(self.NAMESPACE + '.containers.restarts', restart_count, tags)
+
+                for (metric_name, field_name) in [('state', 'state'), ('last_state', 'lastState')]:
+                    c_state = ctr_status.get(field_name, {})
+
+                    for state_name in ['running', 'terminated', 'waiting']:
+                        state_reasons = WHITELISTED_CONTAINER_STATE_REASONS.get(state_name, [])
+                        self._submit_container_state_metric(metric_name, state_name, c_state, state_reasons, tags)
+
+    def _submit_container_state_metric(self, metric_name, state_name, c_state, state_reasons, tags):
+        reason_tags = []
+
+        state_value = c_state.get(state_name)
+        if state_value:
+            reason = state_value.get('reason', '')
+
+            if reason.lower() in state_reasons:
+                reason_tags.append('reason:%s' % (reason))
+
+            gauge_name = '{}.containers.{}.{}'.format(self.NAMESPACE, metric_name, state_name)
```

Comment:
This will be executed even if the reason is not whitelisted, could you please mirror [the ksm logic](https://github.com/DataDog/integrations-core/blob/92f468deadedb6143e2214ca929bebdec8385572/kubernetes_state/datadog_checks/kubernetes_state/kubernetes_state.py#L406-L411), making sure we only submit gauges for whitelisted reasons?

The rationale behind that is that the other states are transient and would increase timeseries churn for little value.


Would be great to add a test with a non-whitelisted value, to make sure we don't have a regression down the road

Reply:
I can change this to submit only for whitelisted reasons.  The current implementation submits a gauge _without_ a reason tag if the reason is not whitelisted.  This allows us to know when containers are in a state, even if we do not have a specific reason.  In the case of `running` (which is likely low value and duplicated elsewhere, but included for completeness), this allows us to submit the gauge with no reason.

Reply:
I do believe having containers with `reason:N/A` will be pretty confusing for most users, although I agree we should open the discussion on this.

Although, as freeze is coming and most of my team is at re:Invent, I'd rather play it safe and merge a mirror of the ksm logic for now, then continue the discussion on how we could improve it. Would that be OK for you?

Reply:
Pushed code to submit only for a valid reason.  This also eliminates the redundant `state.running` metric, which should _significantly_ reduce the number of timeseries submitted from here.

The more I think about it, the more I think you are correct about how to handle these metrics.  Thanks!